### PR TITLE
Feature/mint

### PR DESCRIPTION
**Pull request by Flocqst**
Implement Minting and Conversion Ratio to the KSX Vault

## Description
See spec for Minting and Conversion Ratio :
https://github.com/Kwenta/kwenta-state-log/blob/master/kips/kip-123.md

Add tests to ensure the first $KSX is minted by depositing $KWENTA at an initial fixed ratio of 1,000 $KSX per $KWENTA.

Add KSXVaultBootsrap for tests

### CODE REVIEW

File: test/KSXVault.t.sol [hunk 1]
```diff
@@ -1,20 +1,103 @@
...
+    // Converts asset values to shares and deposits assets into the vault
+    function test_vault_deposit() public {
+        uint256 amount = 1 ether;
+        vm.startPrank(alice);
+        depositToken.approve(address(ksxVault), amount);
+        ksxVault.deposit(1 ether, alice);
+        assertEq(ksxVault.balanceOf(alice), amount * (10 ** ksxVault.offset()));
+        assertEq(stakingRewards.stakedBalanceOf(address(ksxVault)), amount);
+        vm.stopPrank();
+    }
```
Comment (jcmonte) [inline test/KSXVault.t.sol:50]: This test checks that Alice is minted the offset amount specified in the contract, but it does not verify that 1 Kwenta = 1000 KSX. Like if you set offset to 0 this test would still pass. 

You should have a separate test that shows 1 KWENTA deposited results in 1000 KWENTA hardcoded. (original business case)

This should also be fuzzed. For your fuzz test you can multiply the initial KWENTA amount by 1000. You're trying to prove that for all amounts of Kwenta the initial ratio is 1:1000.

Comment (jcmonte) [reply test/KSXVault.t.sol:50]: Secondarily, you should have tests that test consecutive actions. For example, you should be testing

1. Deposit 1 KWENTA receive x amount
2. Deposit 2 KWENTA, now what amount do you receive? Test your hypothesis

and

1. Deposit 1 KWENTA as Alice receive x amount
2. Allow some time to pass and staking rewards to accrue in the contract\*
3. Deposit 1 KWENTA as Bob and confirm the expected amount

\* more of an integration test but you need to have this

---

Your goal with test coverage is to test as much branching logic you can think of. [This SupplySchedule.t](https://github.com/Kwenta/token/blob/main/test/local/contracts/SupplySchedule.test.ts) is the gold standard for testing. Note how it even tests contract state at random points in the time in the future (future weeks expected behavior). Take some time to review this and look how many test cases were covered.

File: test/KSXVault.t.sol [hunk 2]
```diff
@@ -1,20 +1,103 @@
...
+        uint256 amount = 1 ether;
+        vm.startPrank(alice);
+        depositToken.approve(address(ksxVault), amount);
+        ksxVault.mint(1 ether, alice);
+        assertEq(ksxVault.balanceOf(alice), amount);
+        assertEq(
+            stakingRewards.stakedBalanceOf(address(ksxVault)),
+            amount / (10 ** ksxVault.offset())
+        );
+        vm.stopPrank();
```
Comment (jcmonte) [inline test/KSXVault.t.sol:65]: You should store 1 ether as a constant `ONE_KWENTA` and the second reference to 1 ether here as `ONE_KSX`. Makes it hard to parse/understand what you're testing when same value

Labeling is important in testing

Even more confusing on line 60 where it looks like you're asserting ksx balance == original kwenta balance, but really you're checking that KSX balance == 1 KSX

File: test/KSXVault.t.sol [hunk 3]
```diff
@@ -1,20 +1,103 @@
...
+    }
+
+    // Withdraws a specified amount of assets from the vault by burning the
+    // equivalent shares
+    function test_withdraw() public {
```
Comment (jcmonte) [inline test/KSXVault.t.sol:70]: Add a test that tests initial conversion ratio withdraw. ie. where Alice tries to withdraw 1 KWENTA and has to deposit 1000 KSX

Comment (jcmonte) [reply test/KSXVault.t.sol:70]: Secondarily you need more negative state tests. For example 

- trying to withdraw with no shares
- trying to withdraw with more shares than you have

and do the same for deposits

A lot of this is already tested on ERC4626 but you modified the offset so you want to confirm that everything still works as intended. Also some modifications to stake/unstake

File: test/KSXVault.t.sol [hunk 4]
```diff
@@ -1,20 +1,103 @@
...
+        depositToken.approve(address(ksxVault), amount);
+        ksxVault.mint(1 ether, alice);
+        assertEq(stakingRewards.stakedBalanceOf(address(ksxVault)), amount / 1000);
+        assertEq(
+            stakingRewards.stakedBalanceOf(address(ksxVault)),
+            amount / (10 ** ksxVault.offset())
+        );
+
+        ksxVault.redeem(amount, alice, alice);
+        assertEq(ksxVault.balanceOf(alice), 0);
```
Comment (jcmonte) [inline test/KSXVault.t.sol:97]: proper labeling once again please

File: test/KSXVault.t.sol [hunk 5]
```diff
@@ -1,20 +1,103 @@
...
+        // assertEq(stakingRewards.stakedBalanceOf(address(ksxVault)), 0);
+        // assertEq(depositToken.balanceOf(alice), 10 ether);
```
Comment (jcmonte) [inline test/KSXVault.t.sol:99]: remove commented code

File: test/KSXVault.t.sol [hunk 6]
```diff
@@ -1,20 +1,103 @@
...
+        depositToken.mint(alice, 10 ether);
+        depositToken.mint(bob, 10 ether);
 
-    function mint(address to, uint256 amount) external {
-        _mint(to, amount);
+        // vm.prank(alice);
+        // depositToken.approve(address(ksxVault), type(uint256).max);
+
+        // vm.prank(bob);
+        // depositToken.approve(address(ksxVault), type(uint256).max);
```
Comment (jcmonte) [inline test/KSXVault.t.sol:28]: unused?

**Review by jcmonte — CHANGES_REQUESTED** [review]
Testing is a bit too light for logic this intricate. Also you should definitely have unit tests **and** integration tests split out. Where the former is testing purely the function effects and the latter is testing deeper integration with functioning StakingRewards.

File: test/KSXVault.t.sol (hunk 3 above)
Comment (jcmonte) [reply test/KSXVault.t.sol:70]: https://github.com/Kwenta/token/blob/c724e9329f4fcc52b306fe81fdb4a640979dcc4c/test/local/contracts/MultipleMerkleDistributor.test.ts#L224-L269

This test file for the MultipleMerkleDistributor is another good study to understand what kind of negative tests I'm talking about. Don't leave stones unturned

Comment (Flocqst) [reply test/KSXVault.t.sol]: Added new extensive tests for redeeming process.

File: test/KSXVault.t.sol (hunk 2 above)
Comment (Flocqst) [reply test/KSXVault.t.sol]: Done.

File: test/KSXVault.t.sol (hunk 1 above)
Comment (Flocqst) [reply test/KSXVault.t.sol]: Added new extensive tests for deposit process.
